File: J9Ch/src/J_7_Two_Pointers/Optional_11/_610_Two_Sum_Difference_equals_to_target_Medium.cpp
/** 610 Two Sum - Difference equals to target
*/


class Solution {
public:
    /*
     * @param nums an array of Integer
     * @param target an integer
     * @return [index1 + 1, index2 + 1] (index1 < index2)
     */
    vector<int> twoSum7(vector<int> &nums, int target) {
        // write your code here
        vector<int> result;
        unordered_map<int, int> hashMap;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            int tmp = nums[i] - target;
            if (hashMap.find(tmp) != hashMap.end()) {
                result.push_back(hashMap[tmp] + 1);
                result.push_back(i + 1);
                return result;
            }

            tmp = nums[i] + target;
            if (hashMap.find(tmp) != hashMap.end()) {
                result.push_back(hashMap[tmp] + 1);
                result.push_back(i + 1);
                return result;
            }

            hashMap[nums[i]] = i;
        }
     }
};

首页联系我们
Copyright © 2013-2017 九章算法 All Rights Reserved.  京ICP备16004690号-1

File: JS9Ch/src/J_7_Two_Pointers/Optional_11/python_cpp/_610_Two_Sum_Difference_equals_to_target_Medium.py
# /** 610 Two Sum - Difference equals to target



class Solution:
    """
    @param nums {int[]} n array of Integer
    @param target {int} an integer
    @return {int[]} [index1 + 1, index2 + 1] (index1 < index2)
    """
    def twoSum7(self, nums, target):
        # Write your code here
        nums = [(num, i) for i, num in enumerate(nums)]
        target = abs(target)
        n, indexs = len(nums), []

        nums = sorted(nums, key=lambda x: x[0])

        j = 0
        for i in xrange(n):
            if i == j:
                j += 1
            while j < n and nums[j][0] - nums[i][0] < target:
                j += 1
            if j < n and nums[j][0] - nums[i][0] == target:
                indexs = [nums[i][1] + 1, nums[j][1] + 1]

        if indexs[0] > indexs[1]:
            indexs[0], indexs[1] = indexs[1], indexs[0]

        return indexs